WIP Introduction Tutorial

jonas-amr opened the pull request:
First design of the tutorial. 
Since this is still WIP, I am open for any dislikes, likes, corrections and even a redesign if this should not be, how the tutorial was inteded. 
Any comments welcome.

jonas-amr committed:
start

Created a basic structure for the tutorial. Some Text has been written.
WIP

jonas-amr committed:
hyperlinks

managed to insert a hyperlink to the Python Tutorial. Including source
Code does not work yet

jonas-amr committed:
tutorial extended + full users guide started

- extended guide to example creation
- inclusion of code
- WIP: general default.sreg created
- new folder "Full Users Guide" that reuses all Tutorial files.
index-file needs to be updated once the tutorial is completed

jonas-amr committed:
GUI + postprocessing

- added chapters GUI and postprocessing to tutorial
- capitalized all chapter titles for consistency
- removed 'full users guide' to suppress errors

jonas-amr committed:
Tutorial: Controller + Model

- replaceable example files
- quoting code parts in lessons about controller & model

jonas-amr committed:
Model, Controller, ControlLoop

- Finished lessons for model and controller (roughly)
- corrected line-numbers in default and main
- rough lesson "ContorlLoop"

jonas-amr commented:
I'm not sure if this is the correct way of creating a pull request (from my specific branch to your develop). If not, please tell me so I can correct it.

jonas-amr committed:
Tutorial: Postprocessing

- new tutorial for post- and metaprocessing
- not perfect yet, first sketch

cklb commented on .gitignore:
```diff
@@ -44,4 +44,13 @@ examples/results/processing*
+pymoskito/examples/empty/__main__.py
+pymoskito/resources/models/PT1Model.pkl
+pymoskito/examples/Neuer Ordner/__main__.py
+requirements.txt
```
Except for the log files, these files should not be added here

cklb commented on docs/tutorial/GUI.rst:
```diff
@@ -0,0 +1,59 @@
+
+(16) Opened diagrams will appear in the same dock as the 'Placeholder'. 
+     Switch between them by clicking on their tab.
+	 
```
Nice!

cklb commented on docs/tutorial/creating_controller.rst:
```diff
@@ -0,0 +1,60 @@
+
+.. _`useful functions`: https://pymoskito.readthedocs.io/en/stable/modules/controltools.html
+
+.. literalinclude:: basic_files/controller_example.py
```
If possible, try to directly use the files from the ballbeam example add the "control" comments like "#import start" and "import end" there. This will save some work hen changes to these files are made.

jonas-amr replied:
I get what you mean. But if you look at these example files closely, you'll see that they're modified (lables). It was the first solution I found, cutting certain lines from source code. If there's a better way to do this, I'll gladly change the tutorial files.

cklb replied:
You are completly right, adding the labels to inlcude the interesting parts of a file is the good way to do it.
The point here is, that you can add these labels directly into the code of the ballbeam example. I see no problem in adding some more comment lines in those files.

cklb commented on docs/tutorial/creating_controller.rst:
```diff
@@ -0,0 +1,60 @@
+As you can guess from the line numbers, there is no restriction
+to the amount of different controllers you can store in this file:
+
+.. literalinclude:: basic_files/controller_example.py
```
see above

cklb commented on docs/tutorial/creating_controller.rst:
```diff
@@ -0,0 +1,60 @@
+
+__ `useful functions`_
+
+.. literalinclude:: basic_files/controller_example.py
```
same here and for the next two inlcudes

cklb commented on docs/tutorial/creating_default_regimes.rst:
```diff
@@ -0,0 +1,24 @@
+
+	pymoskito/docs/tutorial/basic_files/default.sreg
+	
+This file will be loaded by the code in lines 29-33 of the `main file 
```
Use relative links here, for example:
```
:doc:`the main file <creating_main>`.
```

cklb commented on docs/tutorial/creating_main.rst:
```diff
@@ -0,0 +1,25 @@
+
+	__main__.py
+
+For a Quickstart, you can copy the following code, taken from the `'Ball and Beam' example`_ . 
```
Basically all self references in the docs should be relative (see comment above)

cklb commented on docs/tutorial/creating_main.rst:
```diff
@@ -0,0 +1,25 @@
+
+Please note that the highlighted lines 29-33 refer to another file, which is yet to be created, 
+but not essential to the program. You can either delete these lines, or read more about 
+the default file here_.
```
Same here.

cklb commented on docs/tutorial/creating_main.rst:
```diff
@@ -0,0 +1,25 @@
+
+.. _here: https://pymoskito.readthedocs.io/en/stable/Tutorial/creating_default_regimes.html
+
+.. literalinclude:: ../../pymoskito/examples/ballbeam/__main__.py
```
That's the way to go (I think :P)

cklb commented on docs/tutorial/creating_own_example.rst:
```diff
@@ -0,0 +1,27 @@
+you will need to implement at least a model and a controller.
+
+The toolbox offers a lot more, please find the instructions to
+all modules `here`_.
```
see comment about self references

cklb commented on docs/tutorial/index.rst:
```diff
@@ -0,0 +1,19 @@
+========
+
+Welcome to the PyMoskito Tutorial! It is intended to introduce new users to the toolbox. 
+For more detailed descriptions, please see the `Full Users Guide`_ or the `References`_.
```
and yet antoher self ref

cklb commented on docs/tutorial/starting.rst:
```diff
@@ -0,0 +1,22 @@
+========================
```
Maybe this one can be moved outside of the tutorial in some kind of quick'n'dirty getting started or quickstart for the impatient.

cklb commented on docs/tutorial/useful Code-Snippets.txt:
```diff
@@ -0,0 +1,72 @@
+http://docutils.sourceforge.net/docs/user/rst/quickref.html#inline-markup
```
I think this file is not really required

jonas-amr replied:
that was just for myself. I forgot to add it to .gitignore ^^

cklb reviewed (commented):
Really nice work!

I've added some comments on special topics like self-references to other documents in the docs and the duplicate files in `basic_files` folder.

All in all I think the work is evolving in a good direction.

cklb commented:
The commit history may become a little more straight forward if you would rebase this branch onto the current develop branch, since the meteprocessor branch has now been merged.

jonas-amr commented:
Please do not close this Pull Request! The branch became inaccessable for me after renaming!

cklb committed:
Version bump to 0.2.1

jonas-amr commented:
I succesfully reverted the merge of the 'metaprocessor' branch. We have to keep this in mind. The problems of reverting merges can be read here:
https://github.com/git/git/blob/master/Documentation/howto/revert-a-faulty-merge.txt

cklb commented:
I don't now whether this action has made things easier for you. Since this branch did not start on a clean checkout of the develop branch and the develop branch itself has evolved in the mean time I would strongly recommend to rebase onto the current version of the develop branch.
This will result in this branch only containing fatures that belong to the tutorial.
If you should encounter any problems I may also perform the rebase for you.

jonas-amr commented:
That would be nice. I'm not that familiar with rebasing yet and don't want to mess things up

jonas-amr committed:
Tutorial: self references

-all hyperlinks pointing at self now relative links
- deleted redundant files

cklb commented on pymoskito/examples/ballbeam/controller.py:
```diff
@@ -30,7 +30,7 @@ def __init__(self, settings):
         self.K = pymoskito.controltools.char_coefficients(self._settings["poles"])
 
-    def _control(self, time, trajectory_values=None, feedforward_values=None,
+	def _control(self, time, trajectory_values=None, feedforward_values=None,
```
The travis build showed that you introduced a syntax error here.

jonas-amr committed:
Controller BallBeam: Syntax

solved

jonas-amr committed:
Tutorial: self-references 2

- relative links lost during rebase
- removed 'useful code snippets'

jonas-amr committed:
Tutorial: renaming files

shorter file names

jonas-amr committed:
Tutorial: pictures moved to folder

- moved all pictures to the already existing folder

jonas-amr committed:
gitignore corrections

hopefully the last file that got lost during rebase

jonas-amr committed:
Tutorial: new structure

- new users guide as container
- moved control loop to beginning of tutprial
- moved implementation of model as second

jonas-amr committed:
Graphcs control loop

first draft

jonas-amr committed:
Tutorial: Graphics Control Loop

- added graphics showing the schematics of PyMoskito
- added coloured version for tutorial

jonas-amr committed:
Tutorial: rework intro & simPen

- rework chapter 'tutorial'
- rework chapter 'Understanding the Structure'
- rework chapter 'simulating a simple pendulum'

jonas-amr committed:
Tutorial: rework simPen2

- added graphic for pendulum description
- rework init
- rework starting

jonas-amr committed:
Tutorial: picture rodPend

- update picture for simple Pendulum

jonas-amr committed:
Tutorial: upate pic rodPen 2

second try to update the picture, first time seems to have failed

jonas-amr committed:
Tutorial: minimal System

- inserting a system, that works with the minimum that is possible, will
be used for tutorial
- WIP

jonas-amr committed:
Tutorial: rework model

- changed model equations
- rework of model lesson
- new test.sreg to load simple settings for tests

jonas-amr committed:
Tutorial: new model

- update of model equations in doku

jonas-amr committed:
Tutorial: testing the model

- added new lesson, describing first tests
- added screenshots for this lesson
- removed controller from implementation part and put it behind first
tests

jonas-amr committed:
Tutorial: Controller file

- implemented different versions of the linear state feedback controller
- added test regime for this controller

jonas-amr committed:
Tutorial: rework Controller

- removed unnecessary lines from controller file
- correction of linearized model
- rework of Controller lesson

jonas-amr committed:
Tutorial: update drawings

update for drawings of the control loop and the rod pendulum. WIP, since
textext not working yet

jonas-amr committed:
Tutorial: test controller

- short chapter about how to test the controller
- shorter times in test regime
- minor changes to test_model and new_controller

jonas-amr committed:
Tutorial: thinning code

thinner source code for main, model, controller

jonas-amr committed:
tutorial: restructure & more precision

- changes in structure on multiple layers
- introductory texts more precise
- verbal description of model
- fusion of main and starting
- rework of model
- more details in model test

jonas-amr committed:
Tutorial: update drawings

- nicely rendered using texText
- ctrl_loop now with multiple layers for quick comparisons

jonas-amr committed:
Tutorial: rework controller & model, fine tuning

- rewritten article on controller and model
- minor changes to most articles for consistency

jonas-amr committed:
Tutorial: tidy up

removing all files unnecessary for tutorial

jonas-amr committed:
Tutorial: tidy up 2

removed all pictures not used in  tutorial

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -2,86 +2,72 @@
+    :end-before: #register
+    :lineno-match:
+
+Do not forget to link your controller to the toolbox at the 
```
Maybe link is a little bit misleading here. It's more like a registration process that tells the toolbox that this module is available.
However as your comment above already mentions this instruction should be moved to the section `Starting the Program`

cklb commented on docs/tutorial/test_controller.rst:
```diff
@@ -17,4 +20,8 @@ i.e. the control error:
\ No newline at end of file
+if the control loop reacts the way you would have predicted.
+
+Keep in mind that the controller is implemented static.
```
If you want to use this formulation I'd go for "[...] implemented statically."

cklb commented on docs/tutorial/test_controller.rst:
```diff
@@ -17,4 +20,8 @@ i.e. the control error:
+
+Keep in mind that the controller is implemented static.
+The control law does not adapt to changes in the model parameters 
+due to the linearized model used to calculate the controller gain.
```
Using "default" parameters which may differ from the "real" parameters of the model when calculating the feedback is one thing, using a linearized model which is only sufficiently accurate around a certain steady state is another issue.
I think that it's a good idea to mention these problems but the the comment in its current form causes more confusion than clarity.

cklb commented on docs/tutorial/test_model.rst:
```diff
@@ -11,19 +11,32 @@ you will see the interface of the toolbox:
+activates a drop down menu.
+Clicking again presents all eligible options.
+One of these options as an eligible model is the class RodPendulumModel,
+since it was registered and linked to PyMoskito earlier.
```
Like above, maybe go for "One of these options now shows the :py:class:`RodPendulumModel` since it was introduced to the toolbox by registering it earlier.

cklb commented on docs/tutorial/test_model.rst:
```diff
@@ -11,19 +11,32 @@ you will see the interface of the toolbox:
+
+By clicking on the arrow that appeared on the left of :py:data:`Model`,
+all model parameters and the initial state are displayed. 
+These are taken from the public part of the ordered dictionary defined earlier in the model.
```
Maybe something around the lines of "These are taken from the `public_settings` which have been defined earlier.

cklb commented:
I see some really nice progress here.

jonas-amr committed:
Tutorial: fine tuning 1

- added further reading
- renamed new_example -> system
- matrices for linearized system and control law in new_controller
-

jonas-amr committed:
Tutorial: fine tuning 2

some minor changes to several formulations

jonas-amr committed:
Tutorial: fine tuning 3

- screenshots without transparent bars
- minor changes to code and articles

jonas-amr commented:
I tried moving the register commands from controller.py and model.py to __main__.py of the minimalSystem. This did not work, telling me that the model i tried to register was unknown.

jonas-amr committed:
Tutorial: fine tuning 4

moved call of register method from model and controller to main

cklb commented on docs/tutorial/intro.rst:
```diff
@@ -0,0 +1,20 @@
+
+.. image:: ../pictures/ctrl_loop_intro.png
+
+Every block in this diagram represents an abstract class or base class, that PyMoskito offers.
```
Better: Every block in this diagram represents a configurable part of the control loop, that is implemented as a generic base class.

cklb commented on docs/tutorial/intro.rst:
```diff
@@ -0,0 +1,20 @@
+that implemented classes work well within the context of the toolbox.
+
+From the highlighted classes, the trajectory generator and the model mixer are considered reusable,
+therefore PyMoskito provides these classes fully implemented ready to use.
```
"[...] these classes ready to go."

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+in opposite direction as the speed of rotation :math:`\dot{\varphi}`.
+
+The task is to control the position :math:`s` of the cart 
+and the deflection angle :math:`\varphi` of the pendulum.
```
So that the pendulum is stabilized in either its up- or downside configuration.

jonas-amr replied:
That is not entirely correct, since stabilizing the upward position is ignored during this tutorial.

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. image:: ../pictures/rodPendulum.png
+
+The state vector :math:`\boldsymbol{x}` is given as:
```
With the state vector [...] given as

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+        \dot{\varphi}
+    \end{pmatrix} 
+
+The model equations are given as:
```
the model equations are given by

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+    \end{pmatrix} 
+
+
+The position :math:`s` of the cart is the output of the system: 
```
with the cart's position :math:`s` being the output of the system

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+
+.. math::
+
+    y = x_1 = s
```
Period.

The point here is to include the formulas in the sentences to create a more fluent reading experience.

cklb replied:
At the end of the page, a little preview might be a good thing to add, e.g. "In the next step, these equations will be converted into code that can be simulated"

cklb commented on docs/tutorial/starting.rst:
```diff
@@ -0,0 +1,28 @@
+
+To do so, create a file in the same directory as the model and name it::
+
+    __main__.py
```
Concerning brevity the file name may also just be `main.py`

cklb commented on docs/tutorial/starting.rst:
```diff
@@ -0,0 +1,28 @@
+and start the toolbox with the command::
+
+    $ python __main__.py
+
```
Again, a little explanation about what should happen and that the details will be explained in the next section.

cklb commented on docs/tutorial/intro.rst:
```diff
@@ -0,0 +1,20 @@
+On the other hand, the model and the controller are determined by 
+the specific system and have to be implemented to suit your problem.
+
+If you would like to implement one of the not highlighted classes,
```
-> nonhighlighted

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+System: Pendulum on Cart
+============================
+
+A pendulum is fixed on a cart, which can move in the horizontal direction.
```
[...] as it is shown below:
[Figure from line 23]

cklb commented on docs/tutorial/system.rst:
```diff
@@ -0,0 +1,72 @@
+and the deflection angle :math:`\varphi` of the pendulum.
+Actuating variable is the force :math:`F`.
+
+.. image:: ../pictures/rodPendulum.png
```
Better would be something like:
.. figure:: ../pictures/rodPendulum.png
   :scale: 100 %
   :alt: Image of cart pedulum system

   A paragraph holding the caption

cklb commented on docs/tutorial/new_model.rst:
```diff
@@ -0,0 +1,72 @@
+At first, a new class derived from the abstract class 
+:py:class:`~pymoskito.simulation_modules.Model` is necessary.
+Its basic functions will be calculating the state derivatives and the output,
+dependant on the model parameters, the current state and the input values.
```
depending

cklb commented on docs/tutorial/new_model.rst:
```diff
@@ -0,0 +1,72 @@
+    :end-before: #state
+    :lineno-match:
+
+The calculation of the state derivatives takes place in a method,
```
No comma in front of *that*

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+Implementing a Controller
+-------------------------
+
+A controller closes the loop, a class derived from the abstract class
```
Better: To close the loop a controller has to be added. This can easily be done by deriving from the  :py:class:`~pymoskito.simulation_modules.Controller` class.

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+
+A controller closes the loop, a class derived from the abstract class
+:py:class:`~pymoskito.simulation_modules.Controller`.
+Its only function calculates the controller output,
```
Its task is to stabilize the pendulum by calculating a suitable input for the model.

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+Its only function calculates the controller output,
+which acts as the actuating variable for the model.
+
+In this scenario, the controller is linear. 
```
To keep things simple we start with a controller, based on the linearized system which is given by

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+    \begin{pmatrix}
+        1 & 0 & 0 & 0\\
+    \end{pmatrix} 
+    
```
where

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+.. math::
+    
+    z = (M+m)\cdot J - m^{2}l^{2}
+    
```
.

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+    
+    z = (M+m)\cdot J - m^{2}l^{2}
+    
+The linear control law is given with the control gain :math:`\boldsymbol{K}`
```
Add a little explanation for the calculation of K and V. (e.g. name the ackermann formula)

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+.. math::
+    
+    u = -\boldsymbol{K} \boldsymbol{x} + \boldsymbol{V} \boldsymbol{y_d}
+
```
That's where the theory ends and the fun begins. Maybe add a subheading to mark this point?

jonas-amr replied:
I disagree, since it would be the only subheading, but i inserted a sentence with hopefully the same effect.

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+    :lineno-match:
+
+Name your controller and make :py:class:`~pymoskito.simulation_modules.Controller` its base class.
+Create :py:data:`public_settings` like before in your model.
```
[...] as before in [...]

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+feedforward and controller input.
+The parameter :py:data:`**kwargs` is free to be used as needed, 
+in this case it is ignored.
+Since the controller stabilizes the system in [0,0,0,0],
```
Since this controller will be stabilizing the system in the steady state [....] it has to be subtracted to work on the small signal scale.

cklb commented on docs/tutorial/new_controller.rst:
```diff
@@ -0,0 +1,106 @@
+    :start-after: #control
+    :lineno-match:
+
+Finally, import the controller file and register the controller class to PyMoskito
```
Mention that to process is similar to the model registration

cklb commented on docs/tutorial/test_controller.rst:
```diff
@@ -0,0 +1,30 @@
+since the controller gain is calculated from values stored in the controller class.
+You can use this effect to simulate the situation,
+that the controller was designed based on model parameters 
+that differ from the real parameters of the process.
```
Add an outlook that advertises how a fancy visualization can be added and reference the `Adding a visualization` part of the users guide (which is to be created )

jonas-amr replied:
This already happens briefly in 'further reading'. I could still write a more promoting paragraph about it here. What's your opinion?

jonas-amr committed:
Tutorial: fine tuning 5

- reaction on some comments

jonas-amr committed:
Tutorial: fine tuning 6

reaction on comments

jonas-amr committed:
Tutorial: fine tuning 7

rescaling drawing of RodPendulumSystem

jonas-amr committed:
Tutorial: fine tuning 8

- renamed __main__ -> main
- images -> figures with labels and captions
- improvements of reading flow
